[auto-materialize] enable ignoring parent updates based on run tags

sryza opened the pull request:
## Summary & Motivation

Adds the ability to customize the `materialize_on_parent_updated` auto-materialize rule to only care about updates that correspond to parents whose latest materialization has the provided run tag. This is useful if, for example, you don't want to auto-materialize in response to backfills.

#### SkipOnLatestParentMaterializationHasRunTag?

I originally tried implementing this as its own auto-materialize rule: `SkipOnLatestParentMaterializationHasRunTag`. But I had trouble getting the semantics to achieve the desired behavior.  If an asset has multiple parents, then we'd need to decide whether we:
- skip in the case that ALL parents have the run tag 
  - Problem case: both parents and child were originally auto-materialized. One parent then gets backfilled. Other parent does not have backfill tag, but we still want to skip the child.
- skip in the case that ANY parent has the run tag
  - Problem case: both parents and child were originally backfilled. One parent then gets auto-materialized. Other parent has backfill tag, but we still want to auto-materialize the child.


## How I Tested These Changes

sryza commented on python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:
```diff
@@ -341,14 +341,21 @@ def materialize_on_cron(
         asset, this rule will only fire for the most recent partition of the downstream.
+
+        Args:
+            ignore_updates (Optional[LatestMaterializationHasRunTag]): Filter to apply to updated
```
Right now, this feels oddly specific,  but the thinking here is that, as soon as we have another kind of filter to support, we'd type `ignore_updates` to be `AssetPartitionFilter`, and `LatestMaterializationHasRunTag` would be a subclass or `AssetPartitionFilter`.

sryza replied:
> So the initial comment would be that this should be a sequence of filters.

I had been imagining a composite filter could handle this case.

>  `X.parent_updated_with_tag("dagster/auto_materialize")`

Yeah, an alternative here could just be to introduce a `parent_updated_without_run_tag` rule. The end state here that we could end up with a horde of `parent_updated_X` rules. Not sure how bad that is?

> Another API option would be something along the lines of recontextualizing this argument as something more about "targeting" rather than filtering:

Would it help if we called `AssetPartitionFilter` something like `AssetPartitionTarget` or `AssetPartitionSelection` instead? I like the idea of framing it in terms of targeting, but I'm worried about introducing a horde of arguments to `parent_updated`.

A last, crazy option to consider is for `ignore_updates` or whatever we call it to accept an `AutoMaterializeRule`, that applies to the updated parents.

OwenKephart replied:
Hm if we end up with "composite filters" I worry that this would really muddy the waters in terms of how to combine filters vs. how to combine rules. E.g. you could imagine wanting to have "OrCompositeFilter" and "AndCompositeFilter", etc. If we were to encapsulate all of this logic into specific subclasses of something, i'd much prefer if those at least had a single way to be combined (most likely AND).

Looking closer, I think I'd like to shy away from the horde of `parent_updated_X` rules approach -- it just feels a bit ugly / verbose.

So yeah seems like that narrows it down to either:

```python
X.parent_updated(
    latest_run_tag_key="...",
    latest_run_tag_value="...",
    some_other_parameter="...",
)
```

vs.

```python
X.parent_updated(
    extra_conditions=[
        LatestMaterializationHasRunTag(key="...", value="..."),
        SomeOtherCondition("..."),
    ]
)
```

(or a version of that where it's just `extra_condition = CompositeCondition(...)`)

I guess I don't have as strong priors against a larger set of parameters to `X.parent_updated()` (especially if each individual parameter is somewhat rarely used), and so to my eye the first option feels much better.

It avoids adding any additional concepts for the user ("what exactly is a condition/filter/target?"), and also avoids users having to import additional NamedTuple classes when configuring this thing. It's also easier to discover what configuration options exist as you can just scan the list of parameters rather than having to look for any subclass of `AssetPartitionFilter`.

sryza replied:
I think it's likely that we'll ultimately need to support both of:
- "materialize if a specific parent has been updated with a specific tag"
- "materialize if a specific parent has been updated, or any parent has been updated with a specific tag"

While neither of these are likely to be common, my general observation with auto-materialize customization requests recently is basically that someone is going to want to do everything, and being too flexible is more annoying than not being flexible enough.

If we take the above to be true, then we need to offer flexibility in how users combine different targeting schemes: i.e. we can't just assume one of AND and OR.

If that's the case, then I think an approach based on multiple parameters becomes difficult, because it's not clear how the different conditions should be combined. And ditto with a list of conditions. If that's the case, there's still a question over whether the composition happens at the rule level or at the target/filter level:

```python
X.parent_updated(include=SpecificParent("raw_events")) & X.parent_updated(include=HasTag("dagster/auto_materialize"))
```

vs.

```python
X.parent_updated(include=SpecificParent("raw_events") & HasTag("dagster/auto_materialize"))
```

Thoughts on all this?

OwenKephart replied:
> my general observation with auto-materialize customization requests recently is basically that someone is going to want to do everything

Totally agree -- my feeling was that a single rule is implicitly always going to look for the intersection of all specified constraints (prior art: any of our *Filter classes e.g. RunsFilter, EventRecordsFilter, plus the fact that we can put this in the docstring / docs), and if you want the union of some set of constraints, then you can add additional rules to do this. So for your example, you could rewrite these as:

```python
X.parent_updated(parents={"raw_events"}) & X.parent_updated(tag_key="dagster/auto_materialize")
```
and
```python
X.parent_updated(parents={"raw_events"}, tag_key="dagster/auto_materialize")
```

, respectively. The "horde of parameters" approach also has the benefit of providing a single way of doing the first thing, whereas if we allow for composition of these `include` things, then you could equivalently write that as:

```python
X.parent_updated(include=SpecificParent("raw_events") | HasTag("dagster/auto_materialize"))
```

I think we're describing the same logical semantics w/ the same set of capabilities, just with a different API. I still feel fairly strongly that a big list of parameters is easier for a new user to interact with / discover than this new concept though.

sryza replied:
Two ideas more @OwenKephart and I discussed in a huddle:

```python
AssetPartitionsCondition.parent_updated(
    if_parent_matches=AssetPartitionsCondition.latest_materialization_has_run_tag(key="backfill")
)
```

```python
X.parent_updated(
    if_parent_matches=AssetPartitionsFilter(latest_materialization_has_run_tag_key="backfill")
)
```

I'm coming around to @OwenKephart 's concern that a second, independent system of composable filters would be a bit much.

OwenKephart replied:
Re: offline discussion of the potential for something along the lines of:

```python

AssetPartitionsFilter.parent_updated(
    filter_parents=AssetPartitionsFilter.keys("raw_events") & AssetPartitionsFilter.latest_materialization_has_run_tag(
        "dagster/auto_materialize"
    )
)

```

Pros:
- You could definitely construct situations in which you might want to use the `AssetPartitionsFilter.latest_materialization_has_run_tag(...)` to refer to the current asset (i.e. don't materialize me if my last materialization was tagged with "no/more/auto-materialize"). Somewhat rare use case, of course.
- allows arbitrary complexity in selecting parents (even to the point of allowing user-defined rules I suppose)
- this sort of parameter structure could potentially be used in other rules (and generally gives additional flexibility "for free", as you can toss whatever weird combination of rules you want in there without us explicitly adding more parameters)

Cons:
- Complicates the model of what an AssetPartitionsFilter is -- right now we know that this filter always applies to one specific asset, but if we allow for something like AssetPartitionsFilter.keys(), then it might be harder to calculate these things
- Probably a nightmare to represent this in the UI, now that properties of individual filters can be composed of arbitrary recursively-defined filters (which is a completely parallel hierarchy to the other arbitrarily-recursive structure of filter composition)
- A sort of primal fear arises in my heart when imagining exposing something called `AssetPartitionsFilter`, as this general term (which is not scoped to anything auto-materialize related) implies that this is a sort of tool that you might use for any of your asset partition filtering needs. However, much of its calculation actually _does_ depend on the Asset Daemon machinery keeping track of cursors and whatnot. I guess calling it `AutoMaterializeAssetPartitionsFilter` or something along those lines instead might help a bit

Still somewhat on the fence about this -- it does feel like there could be a good chunk of overlap in the set of things we'd want to filter parents on vs. the set of things we'd want to filter children on, e.g. you could imagine `AssetPartitionsFilter.last_partition()` used to either make sure that only the latest partition of a given asset is kicked off (`(some_filter) & AssetPartitionsFilter.last_partition()`) or to ensure that we only update in response to the last partition of an asset being materialized (although maybe that should just be done using partition mappings)

However, the fractal recursive nature thing is pretty scary to me -- you could get the pathological case of:

```
(complicated_stuff) * X.parent_updated(
    parent_filter = (complicated_stuff2) & X.parent_updated(
        parent_filter = (complicated_stuff3) ...    
    )
)
```

OwenKephart replied:
Summarizing my feelings:

```python
AssetPartitionsCondition.parent_updated(
    if_parent_matches=AssetPartitionsCondition.latest_materialization_has_run_tag(key="backfill")
)
```
- Exposes a ton of surface area and forces us to create a general-purpose evaluation engine, which feels scary and overkill for a thing that will only work in the context of the AssetDaemon

```python
X.parent_updated(
    if_parent_matches=AssetPartitionsFilter(latest_materialization_has_run_tag_key="backfill")
)
```
- At least narrows the scope of things that we have to support to the point where we might be able to let users evaluate these things in user-code space if they so desire. Nice that it can be reused across different rules / conditions, e.g. we might want a separate X.all_parents_updated() method which would also take a filter
- We have to be at least somewhat confident this is a good idea, as we're going to need to whitelist_for_serdes it

```python
X.parent_updated(
    latest_materialization_has_run_tag_key="backfill"
)
```
- Messier than the AssetPartitionsFilter if we end up adding a lot of arguments, not as clear how the parameters should combine
- Simplest solution, requires no new concepts

So I'm leaning towards the second option, but somewhat nervous about how general the name is

OwenKephart commented on python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/updated_scenarios/latest_materialization_run_tag_scenarios.py:
```diff
@@ -0,0 +1,79 @@
+from ..base_scenario import run_request
+from .asset_daemon_scenario_states import one_asset_depends_on_two, two_assets_in_sequence
+
+filter_latest_run_tag_key_policy = (
```
probably worth a test on a situation with multiple forbidden tag keys specified here

OwenKephart commented on python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:
```diff
@@ -561,9 +569,66 @@ def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationRe
+    AutoMaterializeRule,
+    NamedTuple(
+        "_MaterializeOnParentUpdatedRule",
+        [("updated_parent_filter", Optional[AutoMaterializeAssetPartitionFilter])],
```
the addition of a new parameter should modify the description property, as users should have some way, via the UI, to understand that this is different from a non-customized rule. Doesn't necessarily need to be super clean, maybe just:

"upstream data has changed since the latest materialization, subject to the filter: 'latest materialization of parent must not have any of the following run tag keys: {...}'."

Pretty ugly, but only shows up if you're in this niche area.

OwenKephart commented on python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:
```diff
@@ -606,16 +716,61 @@ def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationRe
+                asset_partitions_by_updated_parents.keys()
+                | asset_partitions_by_will_update_parents.keys(),
+            )
+        except RuntimeError as e:
```
can this be removed?

sryza commented:
I made a set of updates here:
- Switched to checking for tags in batch to improve performance.
- No allow specifying a tag key without a tag value, because `RunsFilter` doesn't support this.
- Switched from forbidden tags to required tags. This makes the name less weird and will result in scanning fewer runs, in the use cases we've heard about.

I tried to do a dry run, but ran into a storage inconsistency bug that Phil is currently fixing. I'll kick it off again after that fix makes it in.

OwenKephart commented on python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:
```diff
@@ -561,24 +583,104 @@ def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationRe
+        [("latest_run_required_tags", Optional[Mapping[str, str]])],
+    )
+):
+    """A filter that can be applied to an asset partition, during auto-materialize evaluation, and
```
in reality this is applied to a set of asset partitions, not just one

sryza replied:
From the perspective of the person constructing the class, I think it's easiest to think it about as a per-asset partition filter. The fact that it works in batch is an implementation detail. Open to pushback on this of course.

sryza replied:
Ok actually I'm realizing that both `RunsFilter` and `EventRecordsFilter` are plural, so will change to be consistent.

OwenKephart commented on python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:
```diff
@@ -606,16 +716,54 @@ def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationRe
+                if context.will_update_asset_partition(parent):
+                    asset_partitions_by_will_update_parents[parent].add(asset_partition)
+
+        filtered_updated_and_will_update_parents = self.updated_parent_filter.passes(
```
is it significantly more expensive to do this filtering check inside the above loop? (i'd believe you if you said yes, but if not then a bunch of the complexity below here could be cut out as there's a bunch of shuffling and stuff going on)

OwenKephart replied:
ah I see that you said that it does in the description

sryza replied:
yeah was quite slow :(
